### ISSUE TITLE
Light system

### ISSUE THREAD

**Issue by kremius** [griefly/griefly#370]
There should be light sources (bulbs, flashlights, etc) and without them everything should be dark.

**Comment by shelltitan** [comment]
Do you plan to use shaders or just same blacking method as visible tile? As I know ss13 servers try to implement better shadows.

**Comment by kremius** [comment]
It is possible from a technical point of view.
However, SS13 is a tile based game, so it is hard to make a good looking light/shadow system that does not use per-tile rendering. As far as I can remember such attempts were already made a couple years ago, and they did not seem well for me.
So the initial version of the light system definitely will be the old plain per-tile shading. But if there will be a good idea how to make a fancy shader version of the light system then it always can be considered.

**Comment by shelltitan** [comment]
I thought about one similar that rouge(top-down) likes tend to have.

**Comment by ziepeso** [comment]
Such an attempt was made in Europa Station, it worked quite well, if you ask me.

**Comment by kremius** [comment]
@ziepeso 
If it is possible, can you provide some screenshots? I have googled some, but I would like to be sure when discussing.

**Comment by ziepeso** [comment]
![europ 3](https://cloud.githubusercontent.com/assets/7559793/22987642/27b420b0-f3b8-11e6-9b44-56312c281c8f.jpg)
![europ 2](https://cloud.githubusercontent.com/assets/7559793/22987644/291f6612-f3b8-11e6-929c-7f55b31e64b1.jpg)
![europ](https://cloud.githubusercontent.com/assets/7559793/22987646/2a1c3fe0-f3b8-11e6-881f-f189f18ba8bd.jpg)

**Comment by ziepeso** [comment]
I know they are not the best screenshots, but they give some idea of how the lightning system in Europa worked.

**Comment by ziepeso** [comment]
Here are some better ones.
![aurora](https://cloud.githubusercontent.com/assets/7559793/22987804/b398a8ee-f3b8-11e6-86de-7da7fa2283d9.png)
![europ 4](https://cloud.githubusercontent.com/assets/7559793/22987830/d3816ce0-f3b8-11e6-92db-f238f06b11bd.png)
![europ 5](https://cloud.githubusercontent.com/assets/7559793/22987873/deb55da6-f3b8-11e6-829d-e1add2638adb.png)
![europ 6](https://cloud.githubusercontent.com/assets/7559793/22987895/f2740400-f3b8-11e6-8444-3f6afa6a7df5.png)
Some brighter lightning. Ignore the dozed squid monster.

**Comment by Tacolizard** [comment]
Because Griefly doesn't have the kinds of restrictions BYOND had, I think this is a chance to try to improve SS13's graphics in any way possible. this project shouldn't be restricted to looking like an exact ss13 clone. I think a big turn off for a lot of players is probably the wonky looking interface a crappy fx. basically throw as much modern shader technology as you can into this. or don't. my advice doesn't really matter.

**Comment by shelltitan** [comment]
I have the some opinion about making the graphics better just on the other hand there isn't enough people to work on parallel things, like actually remaking the game and improving it at the same time.

**Comment by kremius** [comment]
@ziepeso

Looks much better than other versions that I have seen!
But there is one issue - it is _too_ pretty for SS13. Other sprites & stuff do not really match the cool light system and it looks a little bit off.
It is possible to actually integrate a light system into sprites:
https://www.youtube.com/watch?v=YXvV5IXvMJw
But it will be not easy. 


@Tacolizard
Sure, everyone wants better graphics. But there are multiple reasons not to do it **now**.
1. It will be possible to play the game without new graphics (because there are a lot of old sprites), but if there will be only new graphics without any engine underneath then instead of game there will be just cool pictures. And making cool graphics takes a lot of effort and time, so it is impossible to develop engine & graphics simultaneously. As far as I know, quite a few SS13 remakes attempts have went the graphics way and that did not end well for them.
2. If (or when) the engine will be more or less ready the new graphics can be added quite easily. For example, the light system from the engine point of view is just a set of points with the intensity of light sources. So pretty light shaders will only render it differently. It still will be immense amount of work, but it will be easier to recruit new people when it is possible to show them that it is possible to play the game.
3. The new graphics should be _better_ than the old ones. It may be not easy to come up with a good idea. For example, people often say that current SS13 UI is quite bad. But they do not present any UI sketch, and I don't actually know that they mean by "SS13 ui is bad". Same with a lot of things.

**Comment by ziepeso** [comment]
@kremius 

You're right, firstly doing a simple light system should be top priority. Graphics come last, since old SS13 only evolved from developer graphics quite late into it's lifespan.  Once the main parts of the game are done (Working atmospherics, electricity, enviromental hazard, combat and medical systems) going for a graphics revamp would be viable, but for now, it's shouldn't be a priority.

**Comment by PJB3005** [comment]
Just saying, but europa lights still looks pretty bad because of the lack of proper blurring/shading techniques available if you're using OpenGL proper. Both SS14's and UnityStation's lighting look miles ahead of it.

Don't handicap yourself by using tile based. Seriously.

**Comment by outbools** [comment]
Majority of modern codebases already using /vg/ version of Goonlights which is not tilebased so i don't quite understand why you should handicap yourself with old outdated system which is started to disappear even from BYOND.

the only reason /tg/ doesnt use Europe is perfomance and bugs.

**Comment by PJB3005** [comment]
> which is not tilebased

Yeah, it is, we're just able to control the 4 corners of the tile. My implementation still looks pretty bad and produces clear zig zag patterns around corners.

**Comment by outbools** [comment]
oh right it's just corners
